[로또 - 웹, DB] 철시 미션 제출합니다.

pkch93 opened the pull request:
현재 페이지 스타일링은 거의 하지않고 서버에서 주고 받은 데이터만 보여주고 있습니다.
백엔드 구성에 대해서 써머님께 빨리 피드백 받고싶어서 먼저 리뷰요청을 보냅니다,
스타일링은 PR 보낸 후 빠르게 처리하겠습니다!

현재 제가 구성한 백엔드 구조에 대해 의문점이 있습니다.

먼저 사용자 입력값을 Dto로 바꿔서 DB에 저장함과 동시에 도메인 객체를 만들고 있는데 옳은 방법인지 궁금합니다.
(LottoService 객체에서 입력값을 받아 Dto로 만들어 DB에 저장하고 도메인 객체를 만들어 return하고 있습니다.)

그리고 WinningNumber와 Result가 LottoGame에 대해 1:1관계를 맺고 있어 LottoGameDao에 WinningNumber와 Result 관련 로직도 넣었는데 분리하는게 맞는지 의견을 여쭙고 싶습니다.

많이 부족한 코드입니다,, 많은 지적 부탁드립니다. 감사합니다 :)

pkch93 committed:
docs: README.md

'로또 1단계 - 자동'에서 필요한 기능 목록 추가

pkch93 committed:
feat: LottoNumber.class

로또 번호를 구현한 LottoNumber 클래스 구현

pkch93 committed:
refactor: LottoNumber.class

1~45로 범위가 정해져 있으므로 성능상 이점을 위해 미리 1~45까지의 LottoNumber를
만든 후 매번 새로운 인스턴스를 만드는 것이 아닌 기존의 값을 가져오는 방식으로 변경

pkch93 committed:
refactor: LottoNumber.class

매직넘버 상수 변경 (1 -> LOTTO_START_NUMBER, 45 -> LOTTO_LAST_NUMBER)

pkch93 committed:
feat: Lotto.class

로또 기능 구현

pkch93 committed:
refactor: Lotto.class

Lotto 클래스의 numbers 타입을 List<Integer> -> List<LottoNumber>로 변경

pkch93 committed:
refactor: Lotto.class

매직넘버 상수화 (6 -> LOTTO_SIZE)

pkch93 committed:
feat: WinningNumber

당첨 번호를 관리하는 객체
사용자가 구매한 로또로 당첨 여부 기능

pkch93 committed:
refactor: LottoTest.class

lotto 객체의 hasNumber가 변함에 따라 테스트 리펙토링

pkch93 committed:
feat: Result.class

최종 결과 목록을 가지고 있는 Result 객체 구현
수익률 계산 기능 구현

pkch93 committed:
refactor: Result.class

매직넘버 상수화 (100.0 -> ONE_HUNDRED_PERCENT)

pkch93 committed:
feat: InputView.class, InputValidator.class

구입금액과 당첨 번호를 입력받는 기능 구현

pkch93 committed:
feat: BoughtLottos.class 구현

사용자가 구매 금액을 입력하면 1000 단위로 절삭하여 로또를 만들어내는 로직 구현
- InputView.class 입력 메세지 추가
- 로또를 생성 기능을 담당하는 LottoGenerator 구현

pkch93 committed:
refactor: Lotto.class, LottoNumber.class

- BoughtLottos 구현 과정에서 Integer가 아닌 LottoNumber 리스트를 받아 Lotto
생성하는 로직으로 변경
- Collections.shuffle 메서드를 활용하기 위해 LottoNumber 클래스의 Map 자료구조를
List로 변경
- 변경에 따른 테스트 리펙토링

pkch93 committed:
style: Result.class

생성자에 한줄 띄워쓰기 제거

pkch93 committed:
feat: ResultGenerator.class 구현

사용자가 산 로또와 당첨 번호를 비교하여 최종 결과를 만들어내는 ResultGenerator
구현

pkch93 committed:
refactor: WINNING_NUMBER_DELIMITER 접근제어자 변경

pkch93 committed:
feat: BoughtLottos.class

구매한 로또의 갯수를 알려주는 메서드 countOfBoughtLottos 구현

pkch93 committed:
refactor: Lotto.class

Lotto과 관리하고 있는 List<LottoNumber>를 반환하는 getNumbers() 구현

pkch93 committed:
refactor: Result.class

상수 MIN_PRIZE_NUMBER와 MAX_PRIZE_NUMBER의 접근제어자 private -> public으로
변경
getCountOfPrize를 수익률 계산 뿐 아니라 OutputView에서도 쓸 수 있을 것이라 판단하여
public으로 접근제어자 변경

pkch93 committed:
refactor: Prize

Prize의 일치 갯수에 접근하는 getCountOfNumber() 구현

pkch93 committed:
feat: OutputView 구현

pkch93 committed:
feat: ConsoleUILottoApplication 구현

pkch93 committed:
style: OutputView 포메팅 및 import 제거

pkch93 committed:
refactor: Result의 상수 MIN_PRIZE_NUMBER와 MAX_PRIZE_NUMBER를 Prize로 이동

- Result보다 Prize에 더 걸맞는 상수라고 판단하여 옮김
- Prize에 옮김에 따라 이를 사용하고 있는 WinningNumber, OutputView 수정

pkch93 committed:
fix: LottoNumber.class

45가 유효한 범위가 아니라는 bug 수정

pkch93 committed:
feat: InputView inputBonusBall 구현

보너스 볼 입력 기능 구현

pkch93 committed:
docs: lotto 2단계 미션 요구사항 및 구현 기능 목록 추가

pkch93 committed:
feat: BonusBall.class 구현

- 구현 과정에서 LottoNumber의 상수값 LOTTO_START_NUMBER와 LOTTO_LAST_NUMBER
접근 제어자 변경 (private -> default)

pkch93 committed:
refactor: LottoGenerator.class

Collections.sort를 List.sort로 변경

pkch93 committed:
feat: Prize 2등(보너스볼 일치) 추가

- Prize 변경에 따른 ResultTest 리펙토링

pkch93 committed:
feat: WinningNumber 보너스 볼 추가

- 보너스볼 추가에 따른 당첨 로직 리펙토링

pkch93 committed:
refactor: Result.class

수익률 계산 로직의 for문을 Prize.values()의 foreach로 리펙토링

pkch93 committed:
feat: OutputView.class

- 2등 결과까지 추가하여 출력 기능 구현

pkch93 committed:
feat: 보너스 볼 입력 기능 구현

pkch93 committed:
refactor: InputValidator의 하드코딩 값 상수화

보너스 볼 입력값 검증을 담당하는 inputValidateBonusBall의 하드코딩 값을 상수화
이와 관련하여 LottoNumber의 LOTTO_START_NUMBER와 LOTTO_LAST_NUMBER의 접근
제어자를 public으로 변경

pkch93 committed:
refactor: printPrizeData CamelCase 변경

pkch93 committed:
refactor: enum Prize 안쓰는 상수 제거

pkch93 committed:
docs: 로또 2단계 완로 후 코드 개선사항 추가

pkch93 committed:
refactor: BonusBall.class의 인스턴스 변수 타입을 int -> LottoNumber로 변환

pkch93 committed:
refactor: WinningNumber.class

WinningNumber의 인스턴스 winningNumbers를 WinningLotto로 변경
이를 위해 WinningLotto 구현, WinningLotto는 당첨 번호를 가지고 있으며
lotto의 매칭갯수를 알려주는 역할을 한다.

이과정에서 BonusBall이 로또의 번호에 있는지 여부는 Lotto 클래스의 역할이라고 판단
하여 hasBonusBall 메서드를 Lotto 클래스의 구현

pkch93 committed:
test: WinningLotto, Lotto

로직 변경하면서 테스트 추가

pkch93 committed:
refactor: OutputView.class의 printPrizeData()

pkch93 committed:
docs: 로또 3단계 요구사항 및 기능목록 추가

pkch93 committed:
feat: 3단계 요구사항 반영

사용자 입력에 따른 Exception 추가
수동 입력 기능 구현
수동 구매 로또에 대한 출력 반영

pkch93 committed:
style: 전체 파일 포메팅 및 사용하지 않는 모듈 import 제거

pkch93 committed:
refactor: 리펙토링에 따른 test 코드 수정

pkch93 committed:
refactor: class BoughtLottos

- 하드코딩 값 상수화 (1000 -> BUY_PRICE)
- 가독성을 위해 자동으로 만들어야하는 로또의 갯수를 변수화

pkch93 committed:
refactor: class Lotto

Lotto의 hasNumber 메서드 명을 checkNumber로 변경

pkch93 committed:
refactor: class LottoNumber

LOTTO_START_NUMBER의 값을 0에서 1로 변경

pkch93 committed:
refactor: Lottosgenerator

LottoGenerator를 LottosGenerator로 변환
수동/자동으로 생성할 수 있도록 interface와 구현체 생성
변경에 따른 코드 변경

pkch93 committed:
refactor: enum Prize

Prize의 각 상수값에 보너스 일치 여부에 대한 값 추가

pkch93 committed:
refactor: enum Prize

requiredBonus를 적용했으나 이전 작업물과 로직이 같게 되어 Prize에 추상메서드로
Prize를 판단하도록 구현

pkch93 committed:
refactor: class Money

Money 클래스를 적용하여 buyPrice 관리

pkch93 committed:
style: 전체 코드 포메팅 및 안쓰는 import 모듈 제거

pkch93 committed:
Merge branch 'pkch93' of https://github.com/pkch93/java-lotto-1 into web

pkch93 committed:
refactor: BoughtLottos에서 수동 / 자동 lotto 생성 합침

pkch93 committed:
refactor: 의미 있는 값 상수화

pkch93 committed:
feat: Money validation 추가

pkch93 committed:
refactor: class BoughtLottos

buyLottos 부분에 바로 구현 클래스를 넣기 보다 lottosGenerate라는 메서드를 만들어
받은 generator를 사용할 수 있도록 구현

pkch93 committed:
docs: README.md

4단계 web UI 관련 정리

pkch93 committed:
fix: 제대로 lotto가 생성이 안되는 부분 수정

- buyLottos 부분에서 price 값이 들어가야하는 부분에 수동 생성 횟수를 넣음

pkch93 committed:
refactor: class단에서 예외처리 코드 추가 및 테스트

- Lotto (중복된 번호가 있는 경우: DuplicateLottoException)
- WinningNumber (로또 번호에 보너스 볼 번호가 있는 경우:
WinningLottoContainBonusException)
- BoughtLottos (구입하려는 금액보다 수동 구입하는 로또의 가격이 큰 경우:
ExceedBoughtLottosAboutMoneyException)

pkch93 committed:
feat: class JsonTransformer

output 값을 json 형태로 변환해 주는 JsonTransformer 구현
이때 gson 라이브러리를 사용

pkch93 committed:
feat: routing 구현

pkch93 committed:
feat: Service 객체 생성

Service Layer를 만들어 request로 오는 값을 처리하는 로직 구현
그 과정에서 BoughtLottos의 buyLottos 부분 인자를 int가 아닌 Money 객체로 변경

pkch93 committed:
refactor: BoughtLottos의 인자가 바뀜에 따른 코드 수정

pkch93 committed:
feat: class WebUILottoApplication

정적 파일 경로 추가 ('/static')

pkch93 committed:
feat: class LottoDao, LottoDto 추가

mysql 연동을 위한 의존성 추가
db 접근을 위한 LottoDao와 Lotto의 전송 객체인 LottoDto 추가

pkch93 committed:
docs: README.md - 5단계 DB 적용

pkch93 committed:
refactor: class LottoDao

Dao에 들어있던 Connection 관련 메서드를 DataConnection 클래스로 분리

pkch93 committed:
feat: class LottoGameDao, LottoGameDto

pkch93 committed:
feat: class LottoGameDao 당첨번호 조회 기능 추가

lotto_game과 winning_lotto가 1:1 대응이므로 LottoGameDao에서 조회하도록
설계
DB와 전송하기 편하도록 WinningNumberDto 생성

pkch93 committed:
feat: class LottoGameDao Result 조회 추가

pkch93 committed:
docs: README.md 5단계 추가내용 작성

pkch93 committed:
feat: class WebUILottoApplication

- port 변경 (8080)
- 자동 리로드를 위해 externalStaticFileLocation 경로 변경

pkch93 committed:
feat: class LottoGameDao

당첨번호, 당첨결과 추가 로직 구현

pkch93 committed:
feat: class LottoService

Dao 객체를 통해 비즈니스 로직을 수행하는 LottoService 구현

pkch93 committed:
refactor: class LottoDao

html 뷰에 보여주기 편하게 getLottoNumbers의 리턴 타입을
List<LottoNumber>를 List<Integer>로 변경

pkch93 committed:
feat: class LottoService

결과 화면에서 필요한 정보를 db에서 가져다 주는 findAllDataOfRound 메서드 구현

pkch93 committed:
feat: view 구현

main.html: lotto를 생성하는 페이지
result.html: 해당 라운드의 lotto 당첨 결과를 보여주는 페이지

pkch93 committed:
feat: html view 구현

pkch93 committed:
feat: class LottoService, LottoGameDao

db를 조회하여 현재 round가 몇 라운드인지 조회하는 기능 구현

pkch93 committed:
test: 기존의 깨진 테스트 수정

pkch93 committed:
refactor: delete class JsonTransformer

ajax를 사용하기 위해 Json을 반환해주는 JsonTransformer
단, ajax 활용을 포기하면서 제거

pkch93 committed:
refactor: class WebUILottoApplication

사용하지 않는 코드 제거

pkch93 committed:
test: LottoGameDaoTest 자동 테스트 적용

이를 위해 LottoGameDao에 result와 winningLotto를 제거하는 로직 구현

pkch93 committed:
docs: README.md

현재 페이지 현황 추가

pkch93 committed:
style: 모든 파일 포메팅 및 import 최적화

pkch93 committed:
fix: mysql password 변경

pkch93 committed:
add: schema.sql, data.sql

jihan805 commented on src/main/java/lotto/dao/LottoGameDao.java:
```diff
@@ -0,0 +1,193 @@
+import static lotto.dao.DataConnection.getConnection;
+
+public class LottoGameDao {
+    private static final int WINNING_START_INDEX = 2;
```
이 부분은 테이블 관련 정보가 없어서 추가로 첨부해주시면 다시 보도록 할게요.
Dao의 경우 하나이 저장소인데 저장소별로 나누는 것이 더 깔끔할것 같아요!

jihan805 replied:
추가로 싱글톤과 DAO에 대해 찾아보고 적용해보면 좋을 것 같아요!

pkch93 replied:
http://www.okjsp.net:8080/article/131372
https://okky.kr/article/583814

주신 피드백을 바탕으로 몇가지 글들을 찾아봤습니다. DAO는 stateless 한 객체이기 때문에 싱글톤으로 보통 사용하는 것을 알 수 있었습니다.

이때, 의문점이 그러면 static 메서드로 사용하면 안되나였는데 (stateless하기 때문에,,) 이에 대해서는 아래 링크를 통해 싱글톤이 static에 비해 OOP적인 측면에서 이점이 있기 때문에 Dao에서 싱글톤을 사용함을 알 수 있었습니다.
https://stackoverflow.com/questions/519520/difference-between-static-class-and-singleton-pattern

좋은 피드백 감사합니다! 혹시라도 제가 잘못 깨우친게 있다면 피드백 부탁드리겠습니다 !!

jihan805 commented on src/main/java/lotto/domain/Money.java:
```diff
@@ -1,9 +1,15 @@
     private final int buyPrice;
 
     public Money(final int buyPrice) {
+        if (buyPrice < BoughtLottos.BUY_PRICE) {
```
validation을 각 클래스로 위임하였네요 👍

jihan805 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,135 @@
+
+import static lotto.domain.generator.LottoNumbersGenerator.generateLottoNumbers;
+
+public class LottoService {
```
서비스 분리 👍 
하지만 클래스가 너무 거대한것 같아요!
로또를 구입 및 생성을 담당하는 LottoService와 결과등을 담당하는 LottoResultService등과 같이 나누어 보면 어떨까요?

pkch93 replied:
피드백을 반영하면서 LottoGameService와 ResultService로 분리하였습니다.

LottoGameService에서 담당해야하는 비즈니스 로직이 로또 구입 및 생성, 결과 도출까지는 필요하다고 생각하였습니다.
반면, ResultService는 뷰에서 해당 round의 결과를 보여주는 데 필요한 정보를 반환하도록 하는 역할을 하도록 구현하고자 했습니다.

jihan805 replied:
분리 👍 
현재는 LottoGameService 에서 로또 구입 ~ 결과 도출까지 "모든 것"을 담당하고 있어요!
하지만, 너무 많은 일들을 하고 있기에 분리해도 좋을 것 같아요!

jihan805 commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -0,0 +1,135 @@
+        return resultDto;
+    }
+
+    public Map<String, Object> findAllDataOfRound(int round) throws SQLException {
```
model에 대한 매핑을 서비스에서 하면 어떤 단점이 있을까요?

pkch93 replied:
https://stackoverflow.com/questions/52423080/is-it-good-to-have-model-mapping-and-more-logics-in-a-mvc-controller

위 링크를 찾아보며 문제점을 찾을 수 있었습니다.
서비스에서 다수의 데이터를 매핑하게 되는 경우 다른 요구사항으로 필요한 데이터를 매핑하는 경우 중복된 코드가 발생하게 되는 문제점이 있을 수 있음을 깨달았습니다,

때문에 각각의 필요한 정보를 (response에 필요한 정보) 반환하고 Mapping 단계를 컨트롤러에서 하게 되었습니다.

주신 피드백에 대해 제가 알아보고 깨달은 부분은 위와 같습니다. 혹시라도 잘못된 부분이 있다면 피드백 부탁드리겠습니다!!

jihan805 reviewed (changes requested):
안녕하세요 철시님!
앞단계 피드백까지 반영 잘 해주셨네요 👍 
피드백 남겼으니 확인 부탁드려요 :)

pkch93 committed:
refactor: class DatabaseConnection

의미를 분명하게 하기 위해 DataConnection에서 DatabaseConnection으로 변경

pkch93 committed:
refactor: class LottoDao, class LottoGameDao

singleton 패턴 적용

pkch93 committed:
refactor: class ResultDao

LottoGameDao가 담당하던 Result 테이블과의 db작업을 ResultDao로 추출

pkch93 committed:
refactor: class ResultService

LottoService의 역할 분리

pkch93 committed:
refactor: class Dto

Dto 클래스에 본 도메인 객체를 만들어주는 toEntity() 메서드 추가
WinningNumber dto의 bonusBall을 bonusNumber로 변수명 변경
ResultService의 인스턴스 변수 prize를 prizeResult로 변경

pkch93 committed:
feat: class LottoNumbersGenerator

List<Integer>를 인자로 받아 List<LottoNumber>를 만들어주는 메서드 구현

pkch93 committed:
refactor: class WinningNumberDao

LottoGameDao가 가진 WinningLotto 테이블 접근 관련 기능을 옮김

pkch93 committed:
refactor: class ResultService

Result 관련 비즈니스 로직을 다루는 ResultService 생성
이에 따른 변화 수정 (LottoService, WebUILottoApplication)

pkch93 committed:
test: Dao

Dao 클래스들 조회 테스트 추가

pkch93 committed:
fix: result.html

ResultDto의 인스턴스 변수 prize를 그대로 쓰고 있었으므로 prizeResult로 변경

pkch93 committed:
test: test용 boughtLottos 생성 방식 변경

jihan805 commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -1,18 +1,52 @@
 
 public class WebUILottoApplication {
-    public static void main(String[] args) {
+    public static void main(String[] args) throws SQLException {
```
해당 클래스가 모든 path를 매핑하고 있어요.
로또 게임이 가지는 기능이 커지면 해당 클래스가 너무 복잡할것 같아요!

jihan805 commented on src/main/java/lotto/dao/LottoDao.java:
```diff
@@ -0,0 +1,91 @@
+    private void setIntFrom(final List<LottoDto> lottoDtos, final PreparedStatement pstmt,
+                            final int dtoIndex) throws SQLException {
+        for (int i = BOUGHT_LOTTO_NUMBER_FROM_INDEX; i < BOUGHT_LOTTO_NUMBER_TO_INDEX; i++) {
+            pstmt.setInt(i + 1, lottoDtos.get(dtoIndex).getLottoNumbers().get(i));
```
한줄에 너무 많은 "."을 사용하지 않도록 해요.
(객체지향생활체조원칙을 참고해보면 좋을 것 같아요)

jihan805 reviewed (approved):
안녕하세요 철시님!
피드백 반영 잘하셨어요 💯
간단한 피드백 남겼으니 확인 부탁드려요. 머지 진행할게요 :)
